File: android/examples/fridgeserver/src/main/java/org/iotivity/base/examples/fridgeserver/DoorResource.java
/*
 * //******************************************************************
 * //
 * // Copyright 2015 Intel Corporation.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * //
 * // Licensed under the Apache License, Version 2.0 (the "License");
 * // you may not use this file except in compliance with the License.
 * // You may obtain a copy of the License at
 * //
 * //      http://www.apache.org/licenses/LICENSE-2.0
 * //
 * // Unless required by applicable law or agreed to in writing, software
 * // distributed under the License is distributed on an "AS IS" BASIS,
 * // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * // See the License for the specific language governing permissions and
 * // limitations under the License.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 */

package org.iotivity.base.examples.fridgeserver;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import org.iotivity.base.EntityHandlerResult;
import org.iotivity.base.OcException;
import org.iotivity.base.OcPlatform;
import org.iotivity.base.OcRepresentation;
import org.iotivity.base.OcResourceRequest;
import org.iotivity.base.OcResourceResponse;
import org.iotivity.base.RequestHandlerFlag;
import org.iotivity.base.ResourceProperty;

import java.util.EnumSet;

import base.iotivity.org.examples.message.IMessageLogger;

/**
 * DoorResource
 * <p/>
 * Creates a door resource and performs action based on client requests
 */
public class DoorResource extends Resource implements IMessageLogger {
    private Context mContext;

    private static String TAG = "DoorResource: ";
    private String mSide = StringConstants.LEFT;
    private boolean mOpen;
    private String resourceURI;

    /**
     * Constructor
     *
     * @param side    left or right side of the door
     * @param context to enable sending of broadcast messages to be displayed on the user screen
     */
    DoorResource(String side, Context context) {
        mContext = context;
        mSide = side;

        resourceURI = StringConstants.DOOR_URI + mSide;

        // eventHandler for register doorResource
        OcPlatform.EntityHandler eh = new OcPlatform.EntityHandler() {
            @Override
            public EntityHandlerResult handleEntity(OcResourceRequest ocResourceRequest) {
                // this is where the main logic of DoorResource is handled
                return entityHandler(ocResourceRequest);
            }
        };
        try {
            logMessage(TAG + "RegisterDoorResource " + resourceURI + " : " +
                    StringConstants.RESOURCE_TYPEDOOR + " : " + StringConstants.RESOURCE_INTERFACE);
            mResourceHandle = OcPlatform.registerResource(resourceURI,
                    StringConstants.RESOURCE_TYPEDOOR, StringConstants.RESOURCE_INTERFACE,
                    eh, EnumSet.of(ResourceProperty.DISCOVERABLE));
        } catch (OcException e) {
            logMessage(TAG + "DoorResource registerResource error: " + e.getMessage());
            Log.e(TAG, e.getMessage());
        }
    }

    /**
     * updates the current value of the door resource
     *
     * @return door representation
     */
    private void updateRepresentationValues() {
        try {
            mRepresentation.setValue(StringConstants.SIDE, mSide);
            mRepresentation.setValue(StringConstants.OPEN, mOpen);
            mRepresentation.setValue(StringConstants.DEVICE_NAME,
                    "Intel Powered 2 door, 1 light refrigerator");
        } catch (OcException e) {
            Log.e(TAG, e.getMessage());
        }
    }

    /**
     * update the OPEN value of doorResource (door is open/ closed)
     *
     * @param representation get current state of door
     */
    private void put(OcRepresentation representation) {
        try {
            mOpen = representation.getValue(StringConstants.OPEN);
        } catch (OcException e) {
            Log.e(TAG, e.getMessage());
        }
        // Note, we won't let the user change the door side!
    }

    /**
     * this is the main method which handles different incoming requests appropriately.
     *
     * @param request OcResourceRequest from the client
     * @return EntityHandlerResult depending on whether the request was handled successfully or not
     */
    private EntityHandlerResult entityHandler(OcResourceRequest request) {
        EntityHandlerResult result = EntityHandlerResult.ERROR;
        if (null != request) {
            try {
                if (request.getRequestHandlerFlagSet().contains(RequestHandlerFlag.REQUEST)) {
                    OcResourceResponse response = new OcResourceResponse();
                    response.setRequestHandle(request.getRequestHandle());
                    response.setResourceHandle(request.getResourceHandle());

                    switch (request.getRequestType()) {
                        case GET:
                            response.setErrorCode(StringConstants.OK);
                            updateRepresentationValues();
                            response.setResourceRepresentation(mRepresentation);
                            response.setResponseResult(EntityHandlerResult.OK);
                            OcPlatform.sendResponse(response);
                            break;
                        case PUT:
                            response.setErrorCode(StringConstants.OK);
                            put(request.getResourceRepresentation());
                            updateRepresentationValues();
                            response.setResourceRepresentation(mRepresentation);
                            response.setResponseResult(EntityHandlerResult.OK);
                            OcPlatform.sendResponse(response);
                            break;
                        case DELETE:
                            response.setResponseResult(EntityHandlerResult.RESOURCE_DELETED);
                            response.setErrorCode(204);
                            OcPlatform.sendResponse(response);
                            break;
                    }
                    result = EntityHandlerResult.OK;
                }
            } catch (OcException e) {
                logMessage(TAG + e.getMessage());
                Log.e(TAG, e.getMessage());
                return EntityHandlerResult.ERROR;
            }
        }
        return result;
    }

    @Override
    public void logMessage(String msg) {
        logMsg(msg);
        if (StringConstants.ENABLE_PRINTING) {
            Log.i(TAG, msg);
        }
    }

    public void logMsg(final String text) {
        Intent intent = new Intent(StringConstants.INTENT);
        intent.putExtra(StringConstants.MESSAGE, text);
        mContext.sendBroadcast(intent);
    }
}

File: service/smarthome-api/server/resource/src/DoorResource.cpp
/* ****************************************************************
 *
 * Copyright 2017 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************/
#include <DoorResource.h>
#include <PropertyBundle.h>
#include <ResourceQuery.h>
#include "logger.h"

#define TAG "SH_SERVER_DOOR_RESOURCE"

namespace OIC
{
    namespace Service
    {
        namespace SH
        {
            const static std::string DOOR_URI = "/door";
            const static std::string OPEN_STATE_PROP_KEY = "openState";
            const static std::string OPEN_DURATION_PROP_KEY = "openDuration";
            const static std::string OPEN_ALARM_PROP_KEY = "openAlarm";

            namespace OpenState
            {
                const std::string Open = "Open";
                const std::string Closed = "Closed";
            };

            DoorResource::DoorResource() :
                    m_delegate(NULL), SHBaseResource(DOOR_URI,
                            RESOURCE_TYPE::DOOR)
            {
                bool ret = false;
                PropertyBundle bundle;

                // Set string property for openState.
                std::string openState = OpenState::Closed;
                bundle.setValue(OPEN_STATE_PROP_KEY, openState);

                // Set string property for openDuration.
                std::string openDuration;
                bundle.setValue(OPEN_DURATION_PROP_KEY, openDuration);

                // Set bool property for openAlarm.
                bool openAlarm = false;
                bundle.setValue(OPEN_ALARM_PROP_KEY, openAlarm);
                
                setPropertyBundle(bundle);
            }

            DoorResource::~DoorResource()
            {
            }

            void DoorResource::setDelegate(DoorResourceDelegate* delegate)
            {
                OIC_LOG(DEBUG, TAG, "Entered setDelegate");
                this->m_delegate = delegate;

                SHBaseResource::setDelegate(this);
            }

            bool DoorResource::changeOpenState(DoorState state)
            {
                PropertyBundle storedBundle = getPropertyBundle();

                std::string openState;
                if (state == OPEN)
                {
                    openState = OpenState::Open;
                }
                else if (state == CLOSED)
                {
                    openState = OpenState::Closed;
                }
                else
                {
                    return false;
                }

                storedBundle.setValue(OPEN_STATE_PROP_KEY, openState);
                setPropertyBundle(storedBundle);                

                return true;
            }

            DoorState DoorResource::getOpenState()
            {
                PropertyBundle storedBundle = getPropertyBundle();

                std::string openState;
                storedBundle.getValue(OPEN_STATE_PROP_KEY, openState);
                
                DoorState state = (openState == OpenState::Open) ? OPEN : CLOSED;

                return state;
            }

            bool DoorResource::changeOpenAlarmState(OpenAlarmState state)
            {
                PropertyBundle storedBundle = getPropertyBundle();

                bool openAlarm = (state == ACTIVATE) ? true : false;
                storedBundle.setValue(OPEN_ALARM_PROP_KEY, openAlarm);
                setPropertyBundle(storedBundle);
                
                return true;
            }

            OpenAlarmState DoorResource::getOpenAlarmState()
            {
                // Get "openAlarm" property value from bundle.
                bool openAlarm;
                PropertyBundle storedBundle = getPropertyBundle();
                storedBundle.getValue(OPEN_ALARM_PROP_KEY, openAlarm);
                
                if (!openAlarm)
                {
                    return DEACTIVATE;
                }
                return ACTIVATE;
            }


            ResultCode DoorResource::onGet(RequestId requestId, const ResourceQuery& query)
            {
                OIC_LOG(DEBUG, TAG, "Entered onGet");

                PropertyBundle storedBundle = getPropertyBundle();
                bool openState = false;
                storedBundle.getValue(OPEN_STATE_PROP_KEY, openState);
                
                return SUCCESS;
            }

            ResultCode DoorResource::onSet(RequestId requestId, const PropertyBundle& bundle,
                                                 const ResourceQuery& query)
            {
                OIC_LOG(DEBUG, TAG, "Entered onSet");

                std::string openState;
                bundle.getValue(OPEN_STATE_PROP_KEY, openState);

                std::string openDuration;
                bundle.getValue(OPEN_DURATION_PROP_KEY, openDuration);

                ResultCode retCode = FAIL;
                bool openAlarm;
                bundle.getValue(OPEN_ALARM_PROP_KEY, openAlarm);

                if (openAlarm)
                {
                    retCode = this->m_delegate->onActivateOpenAlarm();
                }
                else
                {
                    retCode = this->m_delegate->onDeactivateOpenAlarm();
                }

                if (retCode == SUCCESS)
                {
                    // Update property with the requested value.
                    PropertyBundle storedBundle = getPropertyBundle();
                    storedBundle.setValue(OPEN_ALARM_PROP_KEY, openAlarm);                    
                }

                return retCode;
            }
        }
    }
}
